File: Task1-Water_Jug_Problem.cpp
#include <iostream>
using namespace std;

void waterJug(int x, int y, int target) {
    int a = 0, b = 0;

    while (a != target && b != target) {
        // If first jug is empty, fill it
        if (a == 0) {
            a = x;
            cout << "Fill " << x << "L jug";
        }

        // If second jug is full, empty it
        else if (b == y) {
            b = 0;
            cout << "Empty " << y << "L jug";
        }

        // Pour first jug into second jug
        else {
            int amount = min(a, y - b);

            a = a - amount;
            b = b + amount;

            cout << "Pour " << x << "L -> " << y << "L\n";
        }

        cout << "State: (" << a << ", " << b << ")\n\n";
    }

    cout << "Target reached!";
}

int main() {
    waterJug(5, 7, 4);

    return 0;
}

File: Lab_1/Task1-Water_Jug_Problem.py
def water_jug(x, y, target):
    a = 0
    b = 0

    while a != target and b != target:

        # If first jug is empty, fill it
        if a == 0:
            a = x
            print(f"Fill {x}L jug")

        # If second jug is full, empty it
        elif b == y:
            b = 0
            print(f"Empty {y}L jug")

        # Pour first jug into second jug
        else:
            amount = min(a, y - b)

            a = a - amount
            b = b + amount

            print(f"Pour {x}L -> {y}L")

        print(f"State: ({a}, {b})")
        print()

    print("Target reached!")


water_jug(5, 7, 4)